### ISSUE TITLE
LCS checksum offload and other assists

### ISSUE THREAD

**Issue by ivan-w** [SDL-Hercules-390/hyperion#152]
People,

I found the following list of issues related to LCS IP Assists : 
- LCS has a Query IP Assists command frame but no Set Assists command frame (at least none is defined or known - although I *strongly* believe it's cmd code 0xb3)
- During IP Assists discovery, checksum offload is indicated as supported AND enabled since it is always supported (either by the LCS code or by the TAP driver)
- When determining if the TAP driver can do the offload, it is set on and then off on the TAP device

This leads to this situation : If the TAP device can perform IP/TCP/UDP Checksum offloading, the LCS will report that  it is supported and enabled but the TAP device will be setup NOT to perform checksum offloading. When this occurs and the network stack using the LCS queries IP assists, finds that checksum offloading is supported and enabled, that network stack will forego calculating checksums, but NO ONE WILL either - the packet will go out on the network without any checksum and will be dropped eventually at some point.

The solution is either : 
- To leave checksum offload enabled at the TAP level when it is found to be supported
- To leave the offload enabled when it is checked at the TAP level
- To never use the TAP checksum offload (and always use the in-house version instead - but this leads to the next problem)

Now... If the TAP checksum offload is disabled, you cannot enable the TAP TCP Segment Offload (TSO) feature! It's an intricate trick in how the TUNSETOFFLOAD IOCTL handles the command code (it's a bit naive really in the tuntap linux code)

However, you CANNOT forcibly enable TSO (A network stack that is not TSO aware will barf on receiving frames that are way larger than the LAN packet size) - so without a means to enable it, it is completely useless.

So, for the time being
- I am selecting to enable TAP checksum offload (if possible) (it won't hurt.. at worst, the checksum will be computed multiple times on output and frames with bad CRC blocked on TAP ingress).
- Disable TSO (It may still be checked and reported as supported, but won't be reported as enabled by the LCS not enabled at the TAP level).

Patch to follow

**Comment by ivan-w** [comment]
See 676cc85

**Comment by ivan-w** [comment]
The patch does the following for now : 
- Disables TAP checksum offload (I can't get it to work.. the frame is always sent without a checksum)
- Disable TSO (need network stack participation, and since there are no "SET IP Assist" LCS CMD function, it cannot be implemented)
- It also fixes an obscure bug which could lead LCS to hang forever due to an incorrect START LCS Command.

**Comment by mcisho** [comment]
Hercules' LCS support emulates an Ethernet IBM 3172 (or equivalent) operating at layer 2, i.e. the Hercules guest presents Ethernet frames to, and accepts Ethernet frames from, the emulated device. The Hercules guests TCP/IP stacks DEVICE statement specifies a device type of LCS, and the LINK statement specifies a protocol of ETHERNET, for example
  DEVICE LCS1 LCS BA0 
  LINK ETH1 ETHERNET 0 LCS1  

Available from the mid-1990's the IBM 3172 Model 3 supported the TCP/IP Offload feature (for VM or MVS), which introduced, amongst other benefits, the offloading of checksum processing. However, to make use of the Offload device the VM or MVS TCP/IP stacks DEVICE statement specified a device type of CLAW. The DEVICE statement was followed by at least two LINK statements: one IP LINK statement and one or more API LINK statements. For example, OFF1 is a 3172 model 3 with one LAN adapter, running Offload.
  DEVICE OFF1 CLAW 5A0 TCPIP OS2TCP NONE
  LINK IPLINK1 OFFLOADLINK1 1 OFF1
  LINK OFFLAN1 OFFLOADAPIBROAD 9.67.43.200 OFF1 IPLINK1

Though I have no supporting evidence, I suspect that the Query IP Assists command frame was introduced at the time to allow VM or MVS to determine whether the 3172 had the Offload feature. As the Offload software was either running on the 3172 or it wasn't, and the configuration of the Offload feature was done using the OS/2 Multi-Protocol Transport Services (MPTS) facility, a Set Assists command frame probably wasn't needed.

**Comment by ivan-w** [comment]
Nah nah nah.. the CLAW is not a LCS/3172 ! It's a RS/6000 running AIX with a BMX adapter and all traffic is Layer 3 - and use completely different CCWs and protocols !

I Checked... FAL doesn't even try to issue a QIPAssist Command Frame (but that's the old FAL stack, I have to check with a newer one).

z/Linux (lcs.ko) module does issue a QIPAssist command frame upon activation but with z/Linux I'm running into a freak HSCH/Interrupt race condition (working on it).

Now. The problem with NOT having an activation mechanism is not with checksum offloading (calculating CRC multiple times won't hurt - besides a small performance hit), it's with TCP Segmentation Offload (which generates funky ethernet frames to the network stack which are WAY beyond the size of an actual ethernet frame). So if your network stack isn't TSO aware, it won't work with a network interface which forcibly feeds the stack with TSO aggregated frames - so there MUST be a network stack controlled on/off mechanism - or the feature is unusable. Furthermore, the QIPAssist command reply has a "supported" field and an "enabled" field... So it is obviousy designed to be dynamically controlled.

**Comment by mcisho** [comment]
Yes I know a CLAW is not a LCS/3172, but when the 3172 Offload feature is being used it is.

My references are:-
3172 SNA Communications Program and TCP/IP Offload Feature User's Guide (GC30-3675-02)
TCP/IP V3R2 for MVS: Offloading TCP/IP Processing (SC31-7133-02)
I don't have any documentation for early VM TCP/IP stacks.

Today, only z/OS and zLinux does offloading, and then only with QDIO devices.

**Comment by ivan-w** [comment]
I can't find those documentations, but I am not 100% sure they mean "Offload" in the sense of IP/TCP/UDP Checksum Offload and TCP Segmentation Offload.
For CLAW, the "offload" is simply because it is a Layer 2 to Layer 3 "bridge" (and offloads any Layer 2 processing unto the "Workstation")

**Comment by mcisho** [comment]
From the 3172 Offload Feature User's Guide manual:-
The TCP/IP Offload feature performs header and checksum processing, and fragmentation and reassembly of internet protocol (IP) datagrams. This processing usually is performed on the VM or MVS host. The TCP/IP Offload feature reduces the CPU utilization on the System/390 host by performing some of the processing on the 3172 Model 3.

Akaik there has never been an IBM stack that has supported any kind of offload with LCS.

Would you like me to forward the manuals, they're in .boo format?

**Comment by mcisho** [comment]
PDF, way to modern for Softcopy Reader! The book can be printed to a file in .ps format, but there are some sections that you can't select to be printed. Or maybe that's just Softcopy Reader on Linux?

I just forget about offload if I were you, no other VM stack has or does support it.

**Comment by ivan-w** [comment]
I guess I can figure out a way to read them anyway ;)

But as to not use Checksum Offload with an LCS... Too late, already doing it !

--Ivan

**Comment by wably** [comment]
Ivan,

I have the complete VM FAL 2.1 doc , also in bookmanager format.  If you have a Windows machine, I also have the Softcopy reader installer.   Thus you could install that and then read all .boo files.

Alternately, if I know exactly which subject (specifically) that you want, I might be able to print off a few pages to a .PDF and send that to you.  The softcopy reader does not lend itself to printing very well, but small sections are doable. 

Regards,
Bob

**Comment by mcisho** [comment]
All this chat about checksum offload made me look at what SDL's ctc_lcs.c is actually doing. Before the recent disabling of TAP checksum offload, ctc_lcs.c would set an indicator saying that either tuntap would calculate the checksum, or that ctc_lcs.c would calculate the checksum. Following the disabling, the indicator always says that ctc_lcs.c will calculate the checksum.
As a result, for every single TCP, UDP and now ICMP packet ctc_lcs.c sends to tuntap, ctc_lcs.c re-calculates the checksum. However, as there isn't a stack anywhere (sauf Ivan's) that takes any notice of the LCS checksum offload enabled flag, the stack had already calculated the checksums of every single one of those packets.
ctc_lcs.c and/or netsupp.c need to check the IP and ULP header checksums; if they are zero, the stack has not calculated the checksums, if they non-zero the stack has already calculated the checksums. (They will both be either zero or non-zero, there will never be a mixture; zero is not a valid checksum.)

**Comment by ivan-w** [comment]
Apparently, a checksum of 0 is perfectly valid ! (I can give you a valid IP header that yields a zero checksum).
However, a checksum of X'FFFF' (65535) is impossible !
Take any 2 byte field that is aligned, replace it with it's 1 complement, and you get the same checksum).

So for the time being I'll leave it as-is ! 
![image](https://user-images.githubusercontent.com/1625179/49005489-220cf480-f167-11e8-84b1-542fc4575d58.png)

**Comment by mcisho** [comment]
But I thought the point of the checksum offload work was to improve the overall efficiency of Hercules and its guest; as things stand the complete opposite is true.

OK, zero is valid, but it's still the way a z/OS stack tells an OSD that the OSD needs, or doesn't need, to calculate the checksum. And I can confirm that ICMP packets are not checksum offloaded to an OSD that is doing checksum offload for TCP and UDP packets.

**Comment by ivan-w** [comment]
As things stand, the situation is exactly the same as before for stacks that calculate their own checksum ! (the checksum is ALWAYS calculated at the host level no matter what - be it using the tuntap (which calculates on some kernels a wrong checksum) or under hercules (the load of calculating the sum is identical).
The patch only does something that should have been done anyway (note that calculating the checksum at the host level has a very limited impact on performance anyway).

**Comment by mcisho** [comment]
So the checksum is now being calculated three times for LCS, by the Hercules guest, by Hercules, then by tuntap? Hercules doing the calculation when the guest has already done so seems unnecessary overhead to me, but...

**Comment by ivan-w** [comment]
tuntap doesn't perform checksum calculation if it's disabled (which is the case now since we don't invoke ioctl TUNSETOFFLOAD,TUN_F_CSUM
We can't tell if the guest has performed the calculation but granted, we could check for a value of 0 in which case we would only calculate if the value was never calculated (and set to 0) - or the rare case when it is 0

**Comment by ivan-w** [comment]
How about that ? https://github.com/SDL-Hercules-390/hyperion/commit/649b28ab0687ba651661442f58131d4361cd3567

**Comment by mcisho** [comment]
Just the thing. I must just add that netsupp.c's checksum functions are very inefficient, all those ntons are completely unnecessary (see RFC 1071). And why on earth doesn't is support IPv6? And why on earth only ethernet frames?

**Comment by ivan-w** [comment]
I guess you mean htons and ntohs ? (ntons doesn't make much sense !).. Of course, we could simply remove the calls to ntohs - but the benefit is going to be minimal (in most cases, this results in a change in the instruction used to load a 16 bit value in a register - no additional instruction.. on an Intel host it will just change a "mov" to a "movbe")
IPV6 isn't supported because IPV6 doesn't use checksums (it is deemed the link layer error detection and correction is plenty enough).
Now, of course if you want to support TR or FDDI....

**Comment by mcisho** [comment]
Drat, typo! 

No, I don't want to support TR etc! What I meant was, why isn't the function called with the address of an IP packet, rather than an ethernet frame? Make a context specific function into a universally useful function (in Hercules terms  at least).

Though IPv6's IP layer doesn't use checksums, the ULP still does, so why not have the support? Though that does of course mean your recent change would have to change again to check the ULP header for zero. And who knows, someone might want to add checksum offload to qeth one of these days, which would be much easier if netsupp was sensible.

**Comment by ivan-w** [comment]
It's called with the Ethernet frame because that's how you determine the protocol used underneath ! (IPV4, IPV6, ARP, DLC... and whatnot).. Eventually you need to determine the protocol BEFORE you calculate a checksum !

And ... There are no current network stack that support IPV6 over LCS (although it should be possible).

Qeth could very well use netsupp.c (and it has all the infrastructure needed for this).

Granted the naming of functions in netsupp.c may need refactoring !

**Comment by ivan-w** [comment]
Let's not have EVERY driver/handler have to do the same thing - when it could be consolidated into a single set of routines.

**Comment by mcisho** [comment]
I agree, refactoring/splitting/consolidation is the way to go, providing it doesn't obfuscate (which an awful lot of Hercules does, imho).

**Comment by Fish-Git** [comment]
> although I _strongly_ believe it's cmd code 0xb3)

If there's any more information you can provide on this it would be greatly appreciated!

**Comment by ivan-w** [comment]
Purely speculative.. .0xb2 is the query function.. 0xb4 is the next one being used (for Multicast assist)... so I simply deduced it *MAY* be 0xb3.

**Comment by Fish-Git** [comment]
> The book can be printed to a file in .ps format

That, to me, is much better than `.boo` format!&nbsp;&nbsp;`;-)`

(and besides, I'm sure there's a program somewhere out there that can convert .ps (postscript) format files to .pdf format. They're _very_ similar to one another. Almost identical in fact.

**Comment by Fish-Git** [comment]
> If you have a Windows machine, I also have the Softcopy reader installer. Thus you could install that and then read all .boo files.

Can't speak for Ivan, but I've used IBM's Softcopy reader product and it SUCKS! I myself absolutely refuse to use it.

**Comment by Fish-Git** [comment]
> So the checksum is now being calculated three times for LCS: by the Hercules guest, by Hercules, and then by tuntap! Hercules doing the calculation when the guest has already done so seems unnecessary overhead to me, but...

But that's only because the guest, for whatever reason, is purposely doing what we specifically told it was _completely unnecessary for it to do!_ We _told_ it that checksum offloading _was_ supported _and_, not only that, _**was enabled too!**_

The whole point is, it's _vastly_ more efficient for Hercules to calculate the checksums itself (or to let the hardware do it when possible) than it is for the guest to do it using emulated instructions. If we tell the guest that checksum offloading is available _and_ enabled, but they guest calculates the checksums itself anyway, then that's not our fault! That's the fault of the guest! (which we can do nothing about)

**Comment by wably** [comment]
It does suck.  However, most of the documentation that was produced from the 1980s and 1990s simply does not exist in PDF format, thus requiring the use of the Softcopy reader, even if it sucks.  There simply is no alternative if one needs to reference such documentation.

I've searched the internet in the recent past looking for a conversion tool from .boo to .pdf and I have been unable to find one.  I did find one website where someone had started such a tool, but it was never completed.

**Comment by Fish-Git** [comment]
> We can't tell if the guest has performed the calculation, but granted, we could check for a value of 0, in which case we would only calculate it if the value was never calculated (and set to 0) - or the rare case when it is 0.

This seems like a good idea and is probably the best we can do at this point.

I also see where you recently committed exactly such a change too. Thank you!

**Comment by Fish-Git** [comment]
> I must just add that netsupp.c's checksum functions are very inefficient, all those ntons are completely unnecessary (see RFC 1071).

You are absolutely correct! I will see to it that that gets fixed. Thanks for pointing it out.

**Comment by mcisho** [comment]
> But that's only because the guest, for whatever reason, is purposely doing what we specifically told it was _completely unnecessary for it to do!_ We _told_ it that checksum offloading _was_ supported _and_, not only that, _**was enabled too!**_

Because you speak doesn't mean they listen! As I've said several times, there isn't a stack anywhere that expects an LCS to be able to support offload, so there isn't a stack anywhere (except Ivan's) that bothers to check whether offload is supported or enabled.

Did the first OSA's (i.e. pre OSA-Express) support any offload(s)? They were basically LCS's with another name, and maybe new features. Today, the zLinux znetconf command refers to a 3088/60 as an OSA LCS card.

**Comment by ivan-w** [comment]
Then we either support the 0xb2 QUERY IP Assist support, or we don't !If we do then we HAVE to oblige !

**Comment by Fish-Git** [comment]
> Then we either support the 0xb2 QUERY IP Assist support, or we don't! If we do, then we HAVE to oblige!

I'm not sure what you mean, Ivan. We're _already_ supporting 0xb2 QUERY IP Assist! Did you perhaps mean we should add support for 0xb3 SET IP Assist? (and then only enable those assists that the guest specifically requests us to enable?)

**Comment by Fish-Git** [comment]
> Because you speak doesn't mean they listen!

Which seems to be the case for LCS! (Don't know about OSA as I've never bothered to check. Have you? Do you know if z/OS and/or z/VM and/or other guest(s) are always calculating the checksums themselves _regardless_ of our telling it checksum offloading is enabled?)

**Comment by Fish-Git** [comment]
> * Another is TCP Segmentation Offload... THIS cannot be enabled WITHOUT the IP stack being aware of it.
> 
> If Query IP Assist indicates there is a TCP Segment Offload support - does it - it WILL break the underlying TCP/IP stack if it isn't aware of it!

_Unless..._ of course, Hercules provides its _own_ implementation of TCP Segmentation Offloading(*), like it currently does for Checksum Offloading.

Therefore, yes, you are correct: the host network adapter that Hercules uses _cannot_ (_**must** not!_) have TCP Segmentation Offloading(**) enabled on it or else Hercules networking won't work properly. I ran into this myself on Windows. Checksum Offloading? Fine. We can adapt. But TCP Segmentation Offloading? Nope!

p.s. not sure what you mean by "- does it -". 
___
(*) Which isn't for the faint of heart and so should be considered a no-go!
(**) Sometimes also called [Large Send Offload (LSO)](https://en.wikipedia.org/wiki/Large_send_offload).

**Comment by Fish-Git** [comment]
> > I must just add that netsupp.c's checksum functions are very inefficient, all those ntons are completely unnecessary (see RFC 1071).
> 
> You are absolutely correct! I will see to it that that gets fixed. Thanks for pointing it out.

Commits 15d3d1854b96f92acaf754cdb8122148d78b2726 and ba84eaab1804ba03d90a94e708202d5f0eb24af2 together fix this. Thanks again for keeping me on my toes, Ian.

**Comment by ivan-w** [comment]
LSO/TSO is currently disabled on LCS until there is any method to enable/disable it ;)

**Comment by Fish-Git** [comment]
Am closing this issue since I believe everything discussed has since been resolved. Thank you Ivan and Ian for your thoughtful input on this issue! I found it to be a very interesting and enlightening discussion!
